### PR TITLE
prometheus skeleton

### PR DESCRIPTION
**Pull request by thomasrudin**
PR draft for #8004 

# TODO
* Code style/conventions
* Lua implementation (minetest_engine_* and minetest_mod_* namespace)
* Documentation

### CODE REVIEW

File: src/server.cpp [hunk 1]
```diff
@@ -468,11 +481,14 @@ void Server::step(float dtime)
 		}
 		throw ServerError("AsyncErr: " + async_err);
 	}
+
```
Comment (ClobberXD) [inline src/server.cpp]: Please remove this newline if not created on purpose.

File: src/server.h [hunk 2]
```diff
@@ -661,6 +662,12 @@ class Server : public con::PeerHandler, public MapEventReceiver,
 
 	// ModChannel manager
 	std::unique_ptr<ModChannelMgr> m_modchannel_mgr;
+
+	/*
+	Metrics
+	*/
+	Counter* m_tick_count_metric;
+	Counter* m_async_run_count_metric;
```
Comment (ClobberXD) [inline src/server.h]:
```cpp
Counter *m_tick_count_metric;
Counter *m_async_run_count_metric;
```

File: src/monitoring.h [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+	virtual void observe(const double value) = 0;
+};
+
+class Monitoring {
+public:
+	virtual void start() = 0;
+	virtual void stop() = 0;
+	virtual Gauge* createGauge(const std::string &name, const std::string &help) = 0;
+	virtual Counter* createCounter(const std::string &name, const std::string &help) = 0;
+	virtual Histogram* createHistogram(const std::string &name, const std::string &help, const std::vector<double> buckets) = 0;
```
Comment (ClobberXD) [inline src/monitoring.h]: `Gauge *createGauge( ...`

Needs to be corrected in the next two lines as well.

File: src/server.cpp [hunk 4]
```diff
@@ -405,6 +405,17 @@ void Server::init()
 	m_max_chatmessage_length = g_settings->getU16("chat_message_max_size");
 	m_csm_restriction_flags = g_settings->getU64("csm_restriction_flags");
 	m_csm_restriction_noderange = g_settings->getU32("csm_restriction_noderange");
+
+	// Register metrics
+	m_tick_count_metric = g_monitoring->createCounter(
```
Comment (nerzhul) [inline src/server.cpp]: useless metric but interesting example

File: src/CMakeLists.txt [hunk 5]
```diff
@@ -412,6 +421,7 @@ set(common_SRCS
 	mapsector.cpp
 	metadata.cpp
 	modchannels.cpp
+	monitoring.cpp
```
Comment (nerzhul) [inline src/CMakeLists.txt]: compile that only if USE_PROMETHEUS = true

File: src/monitoring.cpp [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU Lesser General Public License for more details.
+
+You should have received a copy of the GNU Lesser General Public License along
+with this program; if not, write to the Free Software Foundation, Inc.,
+51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+*/
+
+#if ENABLE_PROMETHEUS
```
Comment (nerzhul) [inline src/monitoring.cpp]: unneeded, this should be done at cmake configuration time

**Comment by thomasrudin** [comment]
I'm not done yet. Don't worry... ;)

File: src/server.cpp (hunk 4 above)
Comment (thomasrudin) [reply src/server.cpp]: Yeah, that was a simple test-metric :)

File: src/CMakeLists.txt (hunk 5 above)
Comment (thomasrudin) [reply src/CMakeLists.txt]: That won't work: there is a stub-implementation in there for the monitoring if the prometheus library is not found. It is either that or on every metric an `#if PROMETHEUS #endif` clause

**Comment by thomasrudin** [comment]
Can someone please validate my stub-implementation in `prometheus.cpp` and check if thats the right approach?

File: src/CMakeLists.txt (hunk 5 above)
Comment (nerzhul) [reply src/CMakeLists.txt]: you should have instead an interface for this implementation and create a stub implementation when not enabled and a normal implementation when enabled.

Comment (nerzhul) [reply src/CMakeLists.txt]: and also, replace monitoring -> metrology it's a better wording.

Comment (nerzhul) [reply src/CMakeLists.txt]: inheritance: 
```
class IMetrology { // abstract }
class PrometheusMetrology : public IMetrology {}
class StubMetrology : public IMetrology {}
 ```

Comment (thomasrudin) [reply src/CMakeLists.txt]: I won't rename that to metrology, sounds too odd... :/

**Comment by nerzhul** [comment]
why close this ?

**Comment by thomasrudin** [comment]
I realized discussing position of braces and naming of things is just not my thing..

Keep up the good work though :+1:

**Comment by nerzhul** [comment]
code style can be fixed with clang-format if it's new files :)

**Pull request closed without merging**